File: orcastork/graph/circuit_breaker.py
"""Runtime circuit-breaker — bounds a permitted cycle per session.

The orchestrator records each run of a cyclic operator; once an operator reaches its
``max_cycles`` cap it is *tripped* and the orchestrator stops re-running it, halting the
loop. The counter is per-session per-operator — a fresh breaker per session resets it.
"""

from __future__ import annotations

from collections.abc import Mapping

from ..ids import OperatorId


class CircuitBreaker:
    def __init__(self, caps: Mapping[OperatorId, int]) -> None:
        self._caps = dict(caps)
        self._counts: dict[OperatorId, int] = {}

    def record_run(self, operator_id: OperatorId) -> None:
        self._counts[operator_id] = self._counts.get(operator_id, 0) + 1

    def is_tripped(self, operator_id: OperatorId) -> bool:
        cap = self._caps.get(operator_id)
        return cap is not None and self._counts.get(operator_id, 0) >= cap

    def runs(self, operator_id: OperatorId) -> int:
        return self._counts.get(operator_id, 0)

File: typescript/src/orcastork/graph/circuit_breaker.ts
/**
 * Runtime circuit-breaker — bounds a permitted cycle per session.
 *
 * The orchestrator records each run of a cyclic operator; once an operator reaches its `maxCycles`
 * cap it is *tripped* and the orchestrator stops re-running it, halting the loop. The counter is
 * per-session per-operator — a fresh breaker per session resets it.
 *
 * @module
 */

import type { OperatorId } from '../ids.js';

/** Counts runs per operator and trips the ones that reach their cap. */
export class CircuitBreaker {
  private readonly caps: ReadonlyMap<OperatorId, number>;
  private readonly counts = new Map<OperatorId, number>();

  /**
   * `caps` is the cap of every operator on a cycle, as the orchestrator computed it for THIS
   * session's (possibly pruned) gathering set — a `Map`, or any iterable of entries.
   */
  public constructor(caps: Iterable<readonly [OperatorId, number]>) {
    this.caps = new Map(caps);
  }

  public recordRun(operatorId: OperatorId): void {
    this.counts.set(operatorId, (this.counts.get(operatorId) ?? 0) + 1);
  }

  /** True once a capped operator has run its cap; an uncapped operator never trips. */
  public isTripped(operatorId: OperatorId): boolean {
    const cap = this.caps.get(operatorId);
    return cap !== undefined && (this.counts.get(operatorId) ?? 0) >= cap;
  }

  /** How many times this operator has run in this session. */
  public runs(operatorId: OperatorId): number {
    return this.counts.get(operatorId) ?? 0;
  }
}
